go-controller/pkg/ovn/egressip.go

```diff
@@ -2503,15 +2597,15 @@ func (e *EgressIPController) getNextHop(ni util.NetInfo, egressNodeName, egressI
 	isEgressIPv6 := utilnet.IsIPv6String(egressIP)
 	// NOTE: No need to check if status.node exists or not in the cache, we are calling this function only if it
 	// is present in the nodeZoneState cache. Since we call it with lock on cache, we are safe here.
-	if isLocalZoneEgressNode {
+	if isLocalZoneEgressNode || ni.TopologyType() == types.Layer2Topology {
```

Comment:
i think we need a bunch more comments in this function to the different scenarios, because I think this is complex to understand without it being written down. So, this function is trying to determine next hop. Feel free to comment on the below scenarios.

1. If the local node is egress and the interface is not on the ovn network, we send to mp0 as next hop. This should be true for any topology type.
2. If the local node is egress and the interface is ovn networked, then just allow normal routing out of our node via the GR to breth0. This is the same for all topologies.
3. If the egress node is remote, and the topology is layer 3, then on ovn_cluster_router we reroute to the transit switch IP of the remote node as the next hop. This is true only for layer 3.
4. If the egress node is remote, and the topology is layer 2, then on the GR router we reroute to the remote node's GR join IP as the next hop.

agree? Is your code following this logic?

Reply:
> If the local node is egress and the interface is not on the ovn network, we send to mp0 as next hop. This should be true for any topology type.

Agree

> If the local node is egress and the interface is ovn networked, then just allow normal routing out of our node via the GR to breth0. This is the same for all topologies.

Lets say a pod is local and  egress node is also local, therefore for an LRP attached to the pods gateway router, How is ECMP accomplished? 

> If the egress node is remote, and the topology is layer 3, then on ovn_cluster_router we reroute to the transit switch IP of the remote node as the next hop. This is true only for layer 3.

agree

> If the egress node is remote, and the topology is layer 2, then on the GR router we reroute to the remote node's GR join IP as the next hop.

Its GR switch to router port - theres no join switch.

Reply:
> > If the local node is egress and the interface is ovn networked, then just allow normal routing out of our node via the GR to breth0. This is the same for all topologies.
> 
> Lets say a pod is local and egress node is also local, therefore for an LRP attached to the pods gateway router, How is ECMP accomplished?
> 

For layer 3 we need the LRP on ovn_cluster_router with the next hops thats true. For layer 2 as you pointed out this becomes a problem because we are already on the GR, so adding an LRP with multiple nexthops means we have to include the default gateway next hop. Basically cloning routes that already exist in the routing table. It would be nice to have a feature where the nexthops could be (assuming 100.64.0.5 is another egress node) "local_routing_table,100.64.0.5". OVN would interpret local_routing_table to be the value it decided in the previous lr_route_lookup stage in the pipeline and ecmp between those 2 things. @dceara FYI. Since we dont have a feature like that, and with BGP routes can change dynamically in the GR, we would need to constantly update this LRP whenever routes for this pod update. This might be hard to do. In the meantime we can have this is as a limitation in layer 2 egress IP (file a jira), and not do the LRP with multiple next hops when the selected egress node is local.

> > If the egress node is remote, and the topology is layer 2, then on the GR router we reroute to the remote node's GR join IP as the next hop.
> 
> Its GR router to switch port - theres no join switch for L2.

There's no join switch true, but we still use a 2nd subnet (called the join subnet) for handing out unique IP addresses to the GRs and rerouting to them.

Reply:
>  Basically cloning routes that already exist in the routing table. It would be nice to have a feature where the nexthops could be (assuming 100.64.0.5 is another egress node) "local_routing_table,100.64.0.5".

Yes!! I was thinking something similiar because we  if there are multiple default gateway next hops, we cannot just add them into a reroute because it will no longer be ECMP.

Reply:
just documenting (since my future self will forget anyway)...for now we decided leave the LRP with the next hop as the default gw. In the future with BGP we may need to dynamically update these LRPs @jcaamano . Another alternative may be what @dceara suggested with moving to the transit router.

Reply:
Leaving this here too for more context: https://issues.redhat.com/browse/FDP-872

The suggested transit router topology change for L2 UDNs is:
```
  AZ1       AZ2
  ---       ---

  GR1       GR2
   |         |
   +-- TDR --+   
       |
       TS
       |
POD1 --+--- POD2
```